src/Util/PhpServicesCreator.php

```diff
@@ -0,0 +1,574 @@
+            ? $value['decoration_on_invalid']
+            : null;
+
+        if ($invalid) {
+            $this->addUseStatementIfNecessary('Symfony\Component\DependencyInjection\ContainerInterface');
+            $invalid = 'exception' === $invalid
+                ? ', ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE'
+                : ', ContainerInterface::IGNORE_ON_INVALID_REFERENCE'
+            ;
+        }
+
+        // don't write the current and next arguments if null
+        if (null === $invalid && ', null' === $priority) {
+            $priority = null;
+            $innerName = ', null' === $innerName ? null : $innerName;
+        }
+
+        return $this->tab().'->decorate('.$this->createStringArgument($value['decorates']).$innerName.$priority.$invalid.')';
+    }
+
+    private function createCallMethod($values)
+    {
+        $firstKey = array_key_first($values);
+        $firstValue = $values[$firstKey];
+
+        $value = $this->tab().'->call('.$this->createStringArgument($firstValue).', ['.$this->toString($values[1]).']';
+
+        // ->call() can have 3 arguments
+        if (isset($values[2]) && \is_bool($values[2])) {
+            return $values[2] ? $value.', true)' : $value.', false)';
+        }
+
+        return $value.')';
+    }
+
+    private function createTagMethod($value)
+    {
+        if (!\is_array($value)) {
+            return $this->tab().'->tag('.$this->createStringArgument($value).')';
+        }
+
+        $name = $this->createStringArgument($value['name']);
+
+        if (\is_array($value) && \count($value) < 2) {
+            return $this->tab(2).'->tag('.$name.')';
+        }
+
+        $stringArray = '[';
+        foreach ($value as $subKey => $subValue) {
+            if ('name' !== $subKey) {
+                $stringArray .= $this->createStringArgument($subKey).' => '.$this->toString($subValue).', ';
+            }
+        }
+
+        $arrayTag = rtrim($stringArray, ', ').']';
+
+        return $this->tab().'->tag('.$name.', '.$arrayTag.')';
+    }
+
+    private function toString($value)
+    {
+        if ($value instanceof TaggedValue) {
+            if ('service' === $value->getTag()) {
+                $className = $this->addUseStatementIfNecessary($value->getValue()['class']);
+
+                return 'inline('.$className.')';
+            }
+            $args = \is_string($value->getValue())
+                ? $this->createStringArgument($value->getValue())
+                : $this->createListFromArray($value->getValue());
+
+            return $value->getTag().'('.$args.')';
+        }
+
+        switch (\gettype($value)) {
+            case 'array':
+                $value = $this->createArrayArgument($value);
+                break;
+            case 'string':
+                $value = $this->createStringArgument($value);
+                break;
+            case 'boolean':
+                $value = true === $value ? null : 'false';
+                break;
```

Comment:
Include a `default` that throws an exception. That way, if there is some case we're not considering, it will not fail silently.

Reply:
I'm sure you have a better idea than me
`
                throw new \LogicException(sprintf("Only 'array', 'string' or 'boolean' are valid types. Given %d", \gettype($value)));`

Reply:
It's just for act on the value if necessary. But it's probably a better idea to code this:
`


        switch (\gettype($value)) {
            case 'array':
                return $this->createArrayArgument($value);
                break;
            case 'string':
                return $this->createStringArgument($value);
                break;
            case 'boolean':
                return true === $value ? null : 'false';
                break;
            default:
                return $value;
        }`

Reply:
Check how it is now. Switch is used for know if we don't need to act on the value. But if we don't need. We return the value as it is.